File: 2024/05 May/Find the Highest number/main.py
# Link: https://www.geeksforgeeks.org/problems/find-the-highest-number2259/1

from typing import List


class Solution:

    def findPeakElement(self, a: List[int]) -> int:
        """
        Time Complexity: O(log(n))
        Space Complexity: O(1)
        """
        n = len(a)
        low, high = 0, n - 1

        while low < high:
            mid = low + (high - low) // 2

            if a[mid] < a[mid + 1]:
                low = mid + 1
            else:
                high = mid

        return a[high]


if __name__ == "__main__":
    obj = Solution()
    a = [1, 2, 3, 4, 5, 6, 5, 4, 3, 2, 1]
    print(obj.findPeakElement(a))
    a = [1, 2, 3, 4, 5]
    print(obj.findPeakElement(a))

File: 2024/05 May/Find the Highest number/main.cpp
// Link: https://www.geeksforgeeks.org/problems/find-the-highest-number2259/1

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int findPeakElement(vector<int> &a)
    {

        int n = a.size();
        int low = 0, high = a.size() - 1;
        while (low <= high)
        {
            int mid = low + (high - low) / 2;
            int next = (mid + 1) % n;
            int pre = (mid + n - 1) % n;
            if (a[mid] > a[next] && a[mid] > a[pre])
            {
                return a[mid];
            }
            else if (a[mid] > a[mid - 1])
            {
                low = mid + 1;
            }
            else
                high = mid - 1;
        }
    }
};

int main()
{
    Solution obj;
    vector<int> a = {1, 2, 3, 4, 5, 6, 5, 4, 3, 2, 1};
    cout << obj.findPeakElement(a) << endl;

    a = {1, 2, 3, 4, 5};
    cout << obj.findPeakElement(a) << endl;

    return 0;
}
